File: Java/src/main/java/TicketOfficeService.java
import com.google.gson.Gson;

import java.util.*;
import java.util.stream.Collectors;

public class TicketOfficeService {

    private TrainDataClient trainDataClient;
    private BookingReferenceClient bookingReferenceClient;

    public TicketOfficeService(TrainDataClient trainDataClient, BookingReferenceClient bookingReferenceClient) {
        this.trainDataClient = trainDataClient;
        this.bookingReferenceClient = bookingReferenceClient;
    }

    public String makeReservation(ReservationRequestDto reservationRequest) {
        Reservation reservation = makeReservation(reservationRequest.toDomainModel());

        if (reservation.seats.isEmpty()) {
            return "{\"train_id\": \"" + reservation.trainId + "\", \"booking_reference\": \"\", \"seats\": []}";
        }
        return "{" +
                "\"train_id\": \"" + reservation.trainId + "\", " +
                "\"booking_reference\": \"" + reservation.bookingReference + "\", " +
                "\"seats\": [" + reservation.seats.stream().map(s -> "\"" + s.seatNumber + s.coach + "\"").collect(Collectors.joining(", ")) + "]" +
                "}";
    }

    // FIXME : Trop grosse !
    // FIXME : Retourner une réservation
    public Reservation makeReservation(ReservationRequest reservationRequest) {
        // FIXME : Retourner une topologie
        String topologie = trainDataClient.getTopology(reservationRequest.trainId.toString());

        // FIXME: Déplacer (infra)
        Map.Entry<String, List<Topologie.TopologieSeat>> wagonAvecPlace =
                findWagonAvecPlaces(reservationRequest, topologie);
        List<Seat> siegesReserves = selectSiegesDisponibles(reservationRequest, wagonAvecPlace);

        if (!siegesReserves.isEmpty()) {
            Reservation reservation = new Reservation(
                    reservationRequest.trainId,
                    bookingReferenceClient.generateBookingReference(),
                    siegesReserves
            );
            this.bookingReferenceClient.bookTrain(reservation.trainId.toString(), reservation.bookingReference, reservation.seats);
            return reservation;
        } else {
            return new Reservation(
                    reservationRequest.trainId,
                    "",
                    new ArrayList<>()
            );
        }
    }

    private List<Seat> selectSiegesDisponibles(ReservationRequest reservationRequest, Map.Entry<String, List<Topologie.TopologieSeat>> wagonAvecPlace) {
        List<Seat> siegesReserves = new ArrayList<>();
        if(wagonAvecPlace != null) {
            long limit = reservationRequest.seatCount;
            for (Topologie.TopologieSeat siege : wagonAvecPlace.getValue()) {
                if ("".equals(siege.booking_reference)) {
                    if (limit-- == 0) break;
                    Seat seat = new Seat(siege.coach, siege.seat_number);
                    siegesReserves.add(seat);
                }
            }
        }
        return siegesReserves;
    }

    private Map.Entry<String, List<Topologie.TopologieSeat>> findWagonAvecPlaces(ReservationRequest reservationRequest, String topologie) {
        Map.Entry<String, List<Topologie.TopologieSeat>> wagonAvecPlace = null;
        Map<String, List<Topologie.TopologieSeat>> map = new HashMap<>();
        for (Topologie.TopologieSeat x : new Gson().fromJson(topologie, Topologie.class).seats.values()) {
            map.computeIfAbsent(x.coach, k -> new ArrayList<>()).add(x);
        }
        for (Map.Entry<String, List<Topologie.TopologieSeat>> wagon : map.entrySet()) {
            long siegesDisponibles = 0L;
            for (Topologie.TopologieSeat siege : wagon.getValue()) {
                if ("".equals(siege.booking_reference)) {
                    siegesDisponibles++;
                }
            }
            if (siegesDisponibles >= reservationRequest.seatCount) {
                wagonAvecPlace = wagon;
                break;
            }
        }
        return wagonAvecPlace;
    }
}

File: csharp/src/TicketOfficeService.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using Newtonsoft.Json;

namespace KataTrainReservation
{
    public class TicketOfficeService
    {

        private ITrainDataClient trainDataClient;
        private IBookingReferenceClient bookingReferenceClient;

        public TicketOfficeService(ITrainDataClient trainDataClient, IBookingReferenceClient bookingReferenceClient)
        {
            this.trainDataClient = trainDataClient;
            this.bookingReferenceClient = bookingReferenceClient;
        }

        public String MakeReservation(ReservationRequestDto request)
        {
            var seatsByCoachId = GetTopology(request);
            
            var selectedCoach = SelectCoach(request, seatsByCoachId);
            
            var seats = SelectSeats(request, selectedCoach);

            return BookingSeats(request, seats);
        }

        private Dictionary<string, List<Topologie.TopologieSeat>> GetTopology(ReservationRequestDto request)
        {
            var serializedTopology = trainDataClient.GetTopology(request.TrainId);

            var seatsByCoachId = new Dictionary<string, List<Topologie.TopologieSeat>>();
            foreach (var seat in JsonConvert.DeserializeObject<Topologie>(serializedTopology).seats.Values)
            {
                if (!seatsByCoachId.ContainsKey(seat.coach))
                    seatsByCoachId.Add(seat.coach, new List<Topologie.TopologieSeat>());
                seatsByCoachId[seat.coach].Add(seat);
            }

            return seatsByCoachId;
        }

        private static KeyValuePair<string, List<Topologie.TopologieSeat>> SelectCoach(ReservationRequestDto request, Dictionary<string, List<Topologie.TopologieSeat>> seatsByCoachId)
        {
            var selectedCoach =
                default(KeyValuePair<string, List<Topologie.TopologieSeat>>);
            foreach (var coach in seatsByCoachId)
            {
                var freeSeat = 0L;
                foreach (var seat in coach.Value)
                {
                    if ("".Equals(seat.booking_reference))
                    {
                        freeSeat++;
                    }
                }

                var askedSeatNbInReservation = request.SeatCount;
                if (freeSeat >= askedSeatNbInReservation)
                {
                    selectedCoach = coach;
                    break;
                }
            }

            return selectedCoach;
        }

        private static List<Seat> SelectSeats(ReservationRequestDto request, KeyValuePair<string, List<Topologie.TopologieSeat>> selectedCoach)
        {
            var seats = new List<Seat>();
            if (!selectedCoach.Equals(default(KeyValuePair<string, List<Topologie.TopologieSeat>>)))
            {
                var selectedSeats = new List<Seat>();
                long limit = request.SeatCount;
                foreach (var seat in selectedCoach.Value)
                {
                    if ("".Equals(seat.booking_reference))
                    {
                        if (limit-- == 0) break;
                        var selectedSeat = new Seat(seat.coach, seat.seat_number);
                        selectedSeats.Add(selectedSeat);
                    }
                }

                seats = selectedSeats;
            }

            return seats;
        }

        private string BookingSeats(ReservationRequestDto request, List<Seat> seats)
        {
            if (seats.Any())
            {
                var reservation =
                    new ReservationResponseDto(request.TrainId, seats, bookingReferenceClient.GenerateBookingReference());
                bookingReferenceClient.BookTrain(reservation.TrainId, reservation.BookingId, reservation.Seats);
                return "{" +
                       "\"train_id\": \"" + reservation.TrainId + "\", " +
                       "\"booking_reference\": \"" + reservation.BookingId + "\", " +
                       "\"seats\": [" + String.Join(", ", reservation.Seats.Select(s => "\"" + s.SeatNumber + s.Coach + "\"")) +
                       "]" +
                       "}";
            }
            else
            {
                return "{\"train_id\": \"" + request.TrainId + "\", \"booking_reference\": \"\", \"seats\": []}";
            }
        }
    }
}
